File: Dynamic Programming/acmicpc.net/11055/Main.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

class Main {
    static final String inputFilePath = "./Dynamic Programming/acmicpc.net/11055/sample.txt";
    // static final int mod = 1000000000;
    // static final String inputFilePath = "./# format/sample.txt";

    public static void main(String[] args) throws Exception {
        System.setIn(new FileInputStream(inputFilePath));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        int N = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        int a[] = new int[N + 1];
        int d[][] = new int[N + 1][2];
        for (int i = 1; i <= N; i++) {
            a[i] = Integer.parseInt(st.nextToken());
        }
        d[1][1] = a[1];
        for (int i = 2; i <= N; i++) {
            d[i][0] = Math.max(d[i - 1][0], d[i - 1][1]);
            int maxValue = 0;
            for (int j = 1; j < i; j++) {
                if (a[j] < a[i] && maxValue < d[j][1]) {
                    maxValue = d[j][1];
                }
            }
            d[i][1] = maxValue + a[i];
        }
        System.out.println(Math.max(d[N][0], d[N][1]));
    }
}

File: Dynamic Programming/acmicpc.net/11055/Main.js
const inputFilePath = "/sample.txt";
var input = require("fs")
  .readFileSync(__dirname + inputFilePath)
  // .readFileSync('/dev/stdin')
  .toString()
  .trim()
  .split("\n");
const N = input.shift();

const a = input[0].split(" ").map(item => {
  return parseInt(item);
});
a.unshift(0)

const d = [[]];
d.push([0, a[1]]);

for (let i = 2; i <= N; i++) {
  d.push([]);
  d[i][0] = Math.max(d[i - 1][0], d[i - 1][1]);
  let max = 0;
  for (let j = 1; j <= i; j++) {
    if (a[i] > a[j] && max < d[j][1]) {
      max = d[j][1];
    }
  }
  d[i][1] = max + a[i];
}
console.log(Math.max(d[N][0], d[N][1]));
